gneiss/_model.py

```diff
@@ -0,0 +1,124 @@
+import pickle
+from skbio import TreeNode
+import abc
+from skbio.stats.composition import ilr_inv, closure
+import numpy as np
+import pandas as pd
+
+
+class Model(metaclass=abc.ABCMeta):
+
+    def __init__(self, submodels, basis, tree, balances):
+        """
+        Abstract container for balance models.
+
+        Parameters
+        ----------
+        submodels : list of statsmodels objects
+            List of statsmodels result objects.
+        basis : pd.DataFrame
+            Orthonormal basis in the Aitchison simplex.
+            Row names correspond to the leafs of the tree
+            and the column names correspond to the internal nodes
+            in the tree. If this is not specified, then `project` cannot
+            be enabled in `coefficients` or `predict`.
+        tree : skbio.TreeNode
+            Bifurcating tree that defines `basis`.
+        balances : pd.DataFrame
+            A table of balances where samples are rows and
+            balances are columns. These balances were calculated
+            using `tree`.
+        """
+        # TODO: Make sure that the tree and basis are strictly coupled
+        # this will require the development of methods to convert
+        # back and forth between these methods.
+        self.submodels = submodels
+        self.basis = basis
+
+        # Make all nodes in the tree queriable
+        self._nodes = {n.name: n for n in tree.levelorder()}
+
+        self.tree = tree
+        self.balances = balances
+        self.results = []
+
+    @abc.abstractmethod
+    def fit(self, **kwargs):
+        pass
+
+    @abc.abstractmethod
+    def summary(self):
+        """ Print summary results """
+        pass
+
+    def split_balance(self, balance_name):
+        """ Splits a balance into its log ratio components.
+
+        Parameters
+        ----------
+        node : str
+             Name of internal node in the tree to be retrieved for
+
+        Returns
+        -------
+        pd.DataFrame
+            Dataframe where the first column contains the numerator and the
+            second column contains the deminator of the balance.
+        """
+        node = self.tree.find(balance_name)
+        left = node.children[0]
+        right = node.children[1]
+
+        b = np.expand_dims(self.balances[balance_name].values, axis=1)
+        # need to scale down by the number of children in subtrees
+        # there is an erroneous factor 1/sqrt(2) from the ilr_inv below
+        # so we'll need to remove that.
```

Comment:
Not really - this is just a result of the ilr transform.  Specifically, balances are calculated as follows

b_i = np.sqrt(L*R / (L+R)) * np.log(g(x) / g(y))

Where L is the number of leaves in the right tree, R is the number of leaves in the left tree, x is the set of proportions in the left tree, y is the set of proportions in the right tree and g represents the geometric mean.  What we want is to dividate out the sqrt term so we extract the geometric means within the log.  I think I know a more straightforward way to do this, so I'll be pushing in a PR into #86 shortly.